### PR TITLE
Optimizations of av-store

### LINKED ISSUE #2211: Optimize AV Store subsystem.
`av-store` initial implementation lacks optimizations and would ideally benefit from some kind of benchmarking.

Fixes requested on the  internal channel: We always only store the chunk that belongs to our validator when we store the availability data, while the availability distribution will fetch all chunks later Than we also cache nothing. We always go straight to the db. This results in recalculating the data later. And separately storing each chunk as one transaction in the db.

### PR DESCRIPTION
**Pull request by montekki**
Fixes #2211 

1. Compute and store all chunks at once.
2. Use LruCache for chunks
3. Also add two metrics for the sizes of pruning records.

### CODE REVIEW

File: node/core/av-store/src/lib.rs [hunk 1]
```diff
@@ -315,6 +318,7 @@ impl PartialOrd for ChunkPruningRecord {
 pub struct AvailabilityStoreSubsystem {
 	pruning_config: PruningConfig,
 	inner: Arc<dyn KeyValueDB>,
+	chunks_cache: LruCache<(CandidateHash, u32), ErasureChunk>,
```
Comment (bkchr) [inline node/core/av-store/src/lib.rs:321]: Why is that not per candidate?

Comment (bkchr) [reply node/core/av-store/src/lib.rs:321]: However I think a lru cache is a bad idea.

We should keep the chunks as long as a candidate is seen as active and not included or time out. 

The random size of 32 chunks will fail in the moment we increase the number of parachains/validators.

**Review by bkchr — CHANGES_REQUESTED** [review]

File: node/core/av-store/src/lib.rs (hunk 1 above)
Comment (montekki) [reply node/core/av-store/src/lib.rs]: done. Cache will be cleared on either candidate inclusion or timeout together with pruning.

File: node/core/av-store/src/lib.rs [hunk 2]
```diff
@@ -477,6 +483,10 @@ impl AvailabilityStoreSubsystem {
 		Self {
 			pruning_config,
 			inner,
+			// In testing we would want to test actual deletions from DB,
+			// the values in the cache will prevent that so here cache size is
+			// set to `0`.
```
Comment (bkchr) [inline node/core/av-store/src/lib.rs:488]: Why can you not prune properly in tests to ensure that the cache is empty as well?

Comment (montekki) [reply node/core/av-store/src/lib.rs]: yep, you are right, this is leftover after LRU where the cache and store could diverge in their contents. Updated code to remove this

File: node/core/av-store/src/lib.rs [hunk 3]
```diff
@@ -742,11 +753,11 @@ where
 
 			tx.send(result?).map_err(|_| oneshot::Canceled)?;
 		}
-		StoreChunk { candidate_hash, relay_parent, validator_index, chunk, tx } => {
```
Comment (bkchr) [inline node/core/av-store/src/lib.rs:745]: The `validator_index` can be removed from the message

File: node/core/av-store/src/lib.rs [hunk 4]
```diff
@@ -966,54 +980,57 @@ fn store_available_data(
 }
 
 #[tracing::instrument(level = "trace", skip(subsystem), fields(subsystem = LOG_TARGET))]
-fn store_chunk(
+fn store_chunks(
 	subsystem: &mut AvailabilityStoreSubsystem,
 	candidate_hash: &CandidateHash,
-	_n_validators: u32,
-	chunk: ErasureChunk,
+	chunks: &[ErasureChunk],
```
Comment (bkchr) [inline node/core/av-store/src/lib.rs:986]:
```suggestion
	chunks: Vec<ErasureChunk>,
```

File: node/core/av-store/src/lib.rs [hunk 5]
```diff
@@ -966,54 +980,57 @@ fn store_available_data(
...
-	let mut chunk_pruning = chunk_pruning(&subsystem.inner).unwrap_or_default();
-	let prune_at = PruningDelay::into_the_future(subsystem.pruning_config.keep_stored_block_for)?;
+		let prune_at = PruningDelay::into_the_future(subsystem.pruning_config.keep_stored_block_for)?;
+		if let Some(delay) = prune_at.as_duration() {
+			tx.put_vec(
+				columns::META,
+				&NEXT_CHUNK_PRUNING,
+				NextChunkPruning(delay).encode(),
+			);
+		}
```
Comment (bkchr) [inline node/core/av-store/src/lib.rs:1004]: Why is this done for each chunk? That makes no sense or?

File: node/core/av-store/src/lib.rs [hunk 6]
```diff
@@ -966,54 +980,57 @@ fn store_available_data(
...
 
-	if let Some(delay) = prune_at.as_duration() {
-		tx.put_vec(
-			columns::META,
-			&NEXT_CHUNK_PRUNING,
-			NextChunkPruning(delay).encode(),
-		);
-	}
+		let pruning_record = ChunkPruningRecord {
+			candidate_hash: candidate_hash.clone(),
```
Comment (bkchr) [inline node/core/av-store/src/lib.rs:1007]:
```suggestion
			candidate_hash,
```

File: node/core/av-store/src/lib.rs [hunk 7]
```diff
@@ -966,54 +980,57 @@ fn store_available_data(
...
+		);
+	}
 
 	tx.put_vec(
 		columns::META,
 		&CHUNK_PRUNING_KEY,
 		chunk_pruning.encode(),
 	);
 
+
```
Comment (bkchr) [inline node/core/av-store/src/lib.rs:1033]:
```suggestion
```

File: node/core/av-store/src/lib.rs [hunk 8]
```diff
@@ -966,54 +980,57 @@ fn store_available_data(
...
 ) -> Result<(), Error> {
-	let _timer = subsystem.metrics.time_store_chunk();
+	let _timer = subsystem.metrics.time_store_chunks();
 
 	let mut tx = DBTransaction::new();
+	let mut chunk_pruning = chunk_pruning(&subsystem.inner).unwrap_or_default();
 
-	let dbkey = erasure_chunk_key(candidate_hash, chunk.index);
+	for chunk in chunks {
+		subsystem.chunks_cache.entry(*candidate_hash).or_default().insert(chunk.index, chunk.clone());
```
Comment (bkchr) [inline node/core/av-store/src/lib.rs:995]: This has to be moved at the end of loop and changed to
```suggestion
		subsystem.chunks_cache.entry(*candidate_hash).or_default().extend(chunks.into_iter().map(|c| (c.index, c));
```

**Review by bkchr — APPROVED** [review]
Some last nitpicks, otherwise looks good.

**Comment by montekki** [comment]
bot merge

**Pull request merged**